# Assignment 1

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm

from sqlalchemy import create_engine

In [3]:
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'weatherinszeged'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))
df = pd.read_sql_query('select * from weatherinszeged',con=engine)

# no need for an open connection, as we're only doing a single query
engine.dispose()

In [17]:
# Setting the target varible as Y
Y = df['apparenttemperature'] - df['temperature']
# Features set as X
X = df[['humidity','windspeed']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()
rsquared_orig = results.rsquared
rsquared_adj_orig = results.rsquared_adj

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Sun, 22 Mar 2020   Prob (F-statistic):               0.00
Time:                        22:48:41   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.4381      0.021    115.948      0.000       2.397       2.479
humidity      -3.0292      0.024   -126.479      0.000      -3.076      -2.982
windspeed     -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.267
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- Both R-squared and Adj. R-squared values are 0.288.
- While these are low and it would be nice to see them closer to 1 it is not indicative that the model is not working to represent the data.  Some forms of data are harder to reperesent, lowering these numbers to closer to 0.

In [18]:
# Creating interaction of humidity with windspeed in the dataframe
df['humidity_*_windspeed'] = df['humidity'] * df['windspeed']

# Features set as X
X = df[['humidity','windspeed', 'humidity_*_windspeed']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Sun, 22 Mar 2020   Prob (F-statistic):               0.00
Time:                        22:48:46   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
========================================================================================
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
const                    0.0839      0.033      2.511      0.012       0.018       0.149
humidity                 0.1775      0.043      4.133      0.000       0.093       0.262
windspeed                0.0905      0.002     36.797      0.000       0.086       0.095
humidity_*_windspeed    -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.265
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [20]:
print('Difference R-squared: {}'.format(results.rsquared - rsquared_orig))
print('Difference Adj. R-squaredre: {}'.format(results.rsquared_adj - rsquared_adj_orig))

Difference R-squared: 0.05345597954056158
Difference Adj. R-squaredre: 0.05345025808666104


There is an improvment on the model, R-squared and Adj. R-squared have increased from 0.288 to 0.341, a difference of 0.053.

In [21]:
# Setting the target varible as Y
Y = df['apparenttemperature'] - df['temperature']
# Features set as X
X = df[['humidity','windspeed', 'visibility']]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Sun, 22 Mar 2020   Prob (F-statistic):               0.00
Time:                        22:50:15   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.5756      0.028     56.605      0.000       1.521       1.630
humidity      -2.6066      0.025   -102.784      0.000      -2.656      -2.557
windspeed     -0.1199      0.001   -179.014      0.000      -0.121      -0.119
visibility     0.0540      0.001     46.614      0.000       0.052       0.056
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.282
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                          -0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [22]:
print('Difference R-squared: {}'.format(results.rsquared - rsquared_orig))
print('Difference Adj. R-squaredre: {}'.format(results.rsquared_adj - rsquared_adj_orig))

Difference R-squared: 0.015690779810220823
Difference Adj. R-squaredre: 0.015683883687851008


- R-squared and Adj. R-squared both increased from 0.288 to 0.304 and 0.303 respectively.  R-squared and Adj. R-squared increased 0.016.  There is a larger increase in the R value using the interaction feature.
- The lower the better for both AIC and BIC.  The second model with the interaction feature has the lowest at 3.334e+05(333400).